utils/prometheus/src/sourced.rs

```diff
@@ -0,0 +1,138 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Metrics that are collected from existing sources.
+
+use prometheus::core::{Collector, Desc, Describer, Number, Opts};
+use prometheus::proto;
+use std::{cmp::Ordering, marker::PhantomData};
+
+/// A counter whose values are obtained from an existing source.
+pub type SourcedCounter<S> = SourcedMetric<Counter, S>;
+
+/// A gauge whose values are obtained from an existing source.
+pub type SourcedGauge<S> = SourcedMetric<Gauge, S>;
+
+/// The type of a sourced counter.
```

Comment:
To make sure people won't shoot themselves in the foot, what do you think of something along the lines of:

```suggestion
/// The type of a sourced counter.
///
/// A Prometheus counter is a monotonically increasing counter whose value can only ever increase. This property is not enforced here and instead has to be enforced by the user of this crate.
```

Reply:
I added a comment in https://github.com/paritytech/substrate/pull/6895/commits/d341e21a3f60a163d0012357b9caf589712662e4.